frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -832,6 +858,110 @@ pub mod pallet {
+            let current_era = Self::current_era().unwrap_or(Zero::zero());
+            let mut last_claim_era: EraIndex =
+                Self::contract_last_claimed(&contract_id).unwrap_or(current_era.clone());
+            ensure!(
+                current_era != last_claim_era,
+                Error::<T>::AlreadyClaimedInThisEra
+            );
+
+            // is this first time we claim the contract reward?
+            if last_claim_era.is_zero() {
+                last_claim_era = current_era.clone()
+            }
+
+            // oldest era to start with collecting rewards
+            let last_allowed_era = current_era.saturating_sub(Self::history_depth());
+            let start_from_era = last_claim_era.max(last_allowed_era);
+            if (start_from_era > last_claim_era) {
+                // TODO collect all unclaimed rewards and send to Treasury pallet
+            }
+
+            // for the first claimable era "start_from_era", this storage item must be in place!
+            let mut contract_stake_prev = Self::contract_era_stake(&contract_id, &start_from_era)
+                .ok_or(Error::<T>::UnknownStartStakingData)?;
+
+            // for any era after start_from_era, the ContractEraStake is present only if there
+            // was a change in staking amount. If it is not present we process last recorded ContractEraStake
+            for era in start_from_era..current_era {
+                let total_era = Self::get_era_total(era).ok_or(Error::<T>::UnknownEraReward)?;
+                let contract_stake =
+                    Self::contract_era_stake(&contract_id, era).unwrap_or(contract_stake_prev);
+
+                // this contract's part in whole era reward
+                let contract_part = Perbill::from_rational(
+                    Self::balance_to_u64(total_era.rewards).unwrap_or(1),
+                    Self::balance_to_u64(total_era.staked).unwrap_or(1),
+                );
+
+                let contract_era_reward =
+                    contract_part * Self::balance_to_u64(contract_stake.total).unwrap_or(0);
+
+                // divide reward between stakers and the developer of the contract
+                let contract_staker_reward =
+                    Perbill::from_rational((100 - T::DeveloperRewardPercentage::get()) as u64, 100)
+                        * contract_era_reward;
+                let contract_developer_reward =
+                    Perbill::from_rational(T::DeveloperRewardPercentage::get() as u64, 100)
+                        * contract_era_reward;
+
+                // make payout to the stakers and the developer
+                Self::payout_stakers2(&contract_stake, contract_staker_reward);
+                Self::payout_developer(developer.clone(), contract_developer_reward);
+
+                // store current record in case next era has no record of changed stake amount
+                contract_stake_prev = contract_stake;
```

Comment:
We should definitely improve this later to only store the delta per era, not the complete history.

You have to iterate through everything anyhow so it's not that big of a deal.
E.g. this method wouldn't be changed that much.

We don't have to do it for next week's demo though, just an observation.